src/api/v2/mod.rs

```diff
@@ -217,6 +258,47 @@ mod tests {
 			.collect()
 	}
 
+	#[derive(Clone)]
+	struct MockSubmitter {}
+
+	#[async_trait]
+	impl transactions::Submit for MockSubmitter {
+		async fn submit(&self, _: Transaction) -> anyhow::Result<H256> {
+			Ok(H256::random())
+		}
+
+		fn has_signer(&self) -> bool {
+			true
+		}
+	}
+
+	#[tokio::test]
+	async fn submit_route_bad_request() {
+		let route = super::submit_route(Some(Arc::new(MockSubmitter {})));
+		let response = warp::test::request()
+			.method("POST")
+			.path("/v2/submit")
+			.body(r#"{"data":"dHJhbnooNhY3Rpb24:"}"#)
+			.reply(&route)
+			.await;
+		assert_eq!(response.status(), StatusCode::BAD_REQUEST);
+		assert_eq!(response.body(), "Bad Request: Invalid byte 58, offset 17.");
```

Comment:
I think there is no need to check for exact error message as it might change and break test. Checking for status code sounds like enough.

Reply:
Not sure about that one, we should have informative messages to lower confusion of API users during development, which should be part of a test. This is of course just error from json deserialization and could be fragile, but it might be good enough.

Reply:
Lets keep this one also and decide in the future, since there will be more tests and it will be clearer what they represent (also this is still draft PR lol)

Reply:
We will have informative messages, but there is no way to check reliably for it. Imagine you update `serde_json` and message format will change. You would need to change string again, and it is literally a waste of time which won't give you any benefit.

Reply:
IKR, but my point is that we don't have to test very very very specific behavior, like how the error is printed.

Reply:
Understand, but its important what we communicate to end users in case of different errors, so let just think that way here.

Reply:
Of course, but test != communication. Test is related to testing rather. If you want specific error messages, you can document them in RFC or in docs.

Reply:
I would say that tests actually communicate something, like presence of the bugs. But thats not the point here, point is that I want to confirm that two different inputs gives two different error messages with the same status, and thats impossible without checking error message somehow, what's the thing I'm missing here?

Reply:
would say that tests actually communicate something, like presence of the bugs. But thats not the point here, point is that I want to confirm that two different inputs gives two different error messages with the same status, and thats impossible without checking error message somehow, what's the thing I'm missing here?

Reply:
Of course, they communicate something, but that is the last thing which does it:
1. Actual API which literally dictates you what to do
2. Docs for that API which cover some edge cases and spill the light on some corner cases
3. And only right here are tests

My point is that there is no need to check for exact error message. If it was some kind of library, you could have checked for exact error variant, while discarding some unnecessary stuff. Like:
```rust
enum Error {
  A,
  B,
  C(std::io::Error)
}

fn f(x: i32) -> Result<(), Error>;

#[test]
fn test_f() {
	assert_matches!(f(0), Ok(()));
	assert_matches!(f(-1), Err(C(_)));
}
```

But here you just have literal string, which content still has possibility to change with some version of this or that library. There is no good way to check that this error meaning didn't change. So let's just not do it and check for error code.

Reply:
Right way of implementing it is not exposing error messages from any library to end users. but this is a trade off which needs to be at least controlled somehow.

Reply:
Why not to expose error messages, though? Clearly you can't own everything and clearly `serde_json` know better what is the problem with user's json. My point is that there is no need to test serde, especially in this way

Reply:
Anyway, it looks a bit better now. It is not a critical thing, so LGTM!